### PR TITLE
8275908: Record null_check traps for calls and array_check traps in the interpreter

### PR DESCRIPTION
**Pull request by simonis**
`null_checks` occurring at invoke bytecodes are currently not recorded by the profiler. This leads to unnecessary uncommon traps, deoptimizations and recompilations for exceptions which already occurred before the compilation (i.e. are "hot"). This change fixes the problem in the interpreter.

`array_checks` are currently recorded as `class_checks` in the interpreter and therefore not recognized by the compiler. This again leads to uncommon traps, deoptimizations and recompilations. This change unifies the handling of `array_checks` in the interpreter and compiler and prevents unnecessary recompilation.

The test is a stripped down version of a test which was developed for [JDK-8273563: Improve performance of implicit exceptions with -XX:-OmitStackTraceInFastThrow](https://bugs.openjdk.java.net/browse/JDK-8273563) (still [under review](https://github.com/openjdk/jdk/pull/5488)). It introduces an extension to the Whitebox API to expose the decompile, deopt and trap counters which is also required for testing [JDK-8273563](https://bugs.openjdk.java.net/browse/JDK-8273563). I think (and hope) it will also be helpful for others in the future.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8275908](https://bugs.openjdk.java.net/browse/JDK-8275908): Record null_check traps for calls and array_check traps in the interpreter

### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6541/head:pull/6541` \
`$ git checkout pull/6541`

Update a local copy of the PR: \
`$ git checkout pull/6541` \
`$ git pull https://git.openjdk.java.net/jdk pull/6541/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6541`

View PR using the GUI difftool: \
`$ git pr show -t 6541`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6541.diff">https://git.openjdk.java.net/jdk/pull/6541.diff</a>

</details>

### CODE REVIEW

File: test/lib/sun/hotspot/WhiteBox.java
Comment (chhagedorn) [inline test/lib/sun/hotspot/WhiteBox.java:321]: As this class is marked as `@Deprecated`, do we need to add the methods here as well?

**Review by chhagedorn — COMMENTED** [review]
Otherwise, it looks good to me! But would be good to get a second review for it.

Nice test!

File: src/hotspot/share/opto/parseHelper.cpp
Comment (simonis) [reply src/hotspot/share/opto/parseHelper.cpp:301]: The line was actually deleted by my editor. I first wondered myself, but the file had an extra empty line et the end which I think we discourage. So at the end I think my editor was right :)
But as there remained no other changes in that file except the deleted empty line I agree that it looks strange now and I'll restore the file to its initial state.

File: src/hotspot/share/interpreter/interpreterRuntime.cpp
Comment (simonis) [reply src/hotspot/share/interpreter/interpreterRuntime.cpp:834]: Fixed

File: src/hotspot/share/runtime/deoptimization.hpp
Comment (simonis) [reply src/hotspot/share/runtime/deoptimization.hpp:436]: Fixed

File: src/hotspot/share/opto/graphKit.cpp
Comment (TheRealMDoerr) [inline src/hotspot/share/opto/graphKit.cpp:3342]: better: is_aastore

**Review by TheRealMDoerr — COMMENTED** [review]
Nice change! Looks good to me besides what was already said.

File: test/hotspot/jtreg/compiler/exceptions/OptimizeImplicitExceptions.java
Comment (simonis) [reply test/hotspot/jtreg/compiler/exceptions/OptimizeImplicitExceptions.java:287]: Right :)
Fixed.

File: src/hotspot/share/opto/graphKit.cpp
Comment (simonis) [reply src/hotspot/share/opto/graphKit.cpp:3342]: Fixed both occurrences.

File: test/lib/sun/hotspot/WhiteBox.java
Comment (simonis) [reply test/lib/sun/hotspot/WhiteBox.java:321]: Unfortunately yes :(
I first didn't but there are still tests using it and they'll get warning otherwise. To make matters worse, some of them parse the output, so I gave up and added the methods to `sun/hotspot/WhiteBox.java` as well.

File: test/hotspot/jtreg/compiler/exceptions/OptimizeImplicitExceptions.java
Comment (simonis) [reply test/hotspot/jtreg/compiler/exceptions/OptimizeImplicitExceptions.java:141]: Hm, it was used in JDK-8275908 before this fix. Now, with this fix we don't get any recompiles any more :)

The tests is already quite big so I've added a separate test `test/hotspot/jtreg/compiler/uncommontrap/Decompile.java` to verify the new WhiteBox methods introduced by  this change.

File: test/hotspot/jtreg/compiler/exceptions/OptimizeImplicitExceptions.java
Comment (simonis) [reply test/hotspot/jtreg/compiler/exceptions/OptimizeImplicitExceptions.java:166]: That's a good point, now that the tests have got a little simpler :)

I'll have to add more cases for JDK-8273563 but I think it's just fair to leave it for that change.

**Comment by simonis** [comment]
@TheRealMDoerr, @chhagedorn thanks a lot for the quick reviews.

I hope could address all your concerns and suggestions with my latest push.

@chhagedorn: I'm especially happy that you like the tests. As all too often the effort for a good test is much higher than for the fix itself :)

**Review by chhagedorn — APPROVED** [review]
Thanks for doing the changes, they look good to me!

> @chhagedorn: I'm especially happy that you like the tests. As all too often the effort for a good test is much higher than for the fix itself :)

Yes, I couldn't agree more to that. It's often underestimated how much time that is sometimes needed to come up with a good test. So, I always appreciate the extra effort :)

**Comment by simonis** [comment]
Thanks for the approval @chhagedorn .

I found that the new `Decompile.java` test failed on linux/x86_32. The reason for this is that 32-bit builds don't include JVMCI and without JMVCI the bimorphic inlining trap is called just `bimorphic` (in contrast to `bimorphic_or_optimized_type_check` for JVMCI builds`).

The fix is trivial and I hope your approval is also valid for the latest version :)

**Review by TheRealMDoerr — APPROVED** [review]
Right, non-JVMCI platforms need this fix (also PPC and s390). LGTM, now. We'll retest it.

**Review by chhagedorn — APPROVED** [review]
Good catch! Looks good.

**Comment by simonis** [comment]
Thanks @chhagedorn, @TheRealMDoerr !

**Comment by dholmes-ora** [comment]
The new tests are not written correctly and are causing failures in our tier3 CI runs.

If you set an explicit GC on the @run line then you have to ensure no GC option is passed in when running jtreg, else you get two GC's request and the test fails to run.

https://bugs.openjdk.java.net/browse/JDK-8277878

**Pull request closed without merging**
